### PR TITLE
Test errors on cores

### PR DESCRIPTION
**Pull request by imjfckm**

### CODE REVIEW

**Comment by robertbaldyga** [comment]
retest this please

File: test/functional/tests/fault/test_primary_device_error.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        fio_pid = good_core_fio.run_in_background()
+
+    fio_errors = 0
+    with TestRun.step(
+        "Run fio on error core and check if error count matches with cache stats"
+    ):
+        fio_errors = error_core_fio.run()[0].total_errors()
+
+        if fio_errors == 0:
+            TestRun.fail("No I/O ended with errror!")
```
Comment (rafalste) [inline test/functional/tests/fault/test_primary_device_error.py]: Just a small message typo. Or maybe it's a pirate's errrror? ;)

File: test/functional/tests/conftest.py [hunk 2]
```diff
@@ -118,6 +118,7 @@ def pytest_runtest_teardown():
                 if TestRun.executor.is_active():
                     TestRun.executor.wait_for_connection()
                 Udev.enable()
+                kill_all_io()
```
Comment (micrakow) [inline test/functional/tests/conftest.py]: Does this commit update framework as well or is this line the only chagne?

File: test/functional/tests/fault/test_primary_device_error.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+    with TestRun.step("Starting and configuring cache"):
+        cache = casadm.start_cache(
+            cache_device.partitions[0],
+            cache_mode=cache_mode,
+            cache_line_size=cache_line_size,
+            force=True,
+        )
+        result = cache.set_seq_cutoff_policy(SeqCutOffPolicy.never)
+        if result.exit_code:
```
Comment (micrakow) [inline test/functional/tests/fault/test_primary_device_error.py]: There's no need for checking err code, if it is not 0 exception will be raised.

File: test/functional/tests/fault/test_primary_device_error.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+            TestRun.LOGGER.exception("Couldn't set seq cutoff policy")
+        result = cache.set_cleaning_policy(CleaningPolicy.nop)
+        if result.exit_code:
```
Comment (micrakow) [inline test/functional/tests/fault/test_primary_device_error.py]: There's no need for checking err code, if it is not 0 exception will be raised.

File: test/functional/tests/fault/test_primary_device_error.py (hunk 1 above)
Comment (imjfckm) [reply test/functional/tests/fault/test_primary_device_error.py]: Fixed

File: test/functional/tests/conftest.py (hunk 2 above)
Comment (imjfckm) [reply test/functional/tests/conftest.py]: After rebase artifact, modified the message to reflect what's actually happening

File: test/functional/tests/fault/test_primary_device_error.py (hunk 3 above)
Comment (imjfckm) [reply test/functional/tests/fault/test_primary_device_error.py]: Fixed

File: test/functional/tests/fault/test_primary_device_error.py (hunk 4 above)
Comment (imjfckm) [reply test/functional/tests/fault/test_primary_device_error.py]: Fixed

File: test/functional/tests/fault/test_primary_device_error.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    )
+    if io_dir == ReadWrite.randwrite:
+        error_core_fio = error_core_fio.verify_pattern().verify(VerifyMethod.pattern)
+
+    with TestRun.step("Run fio on core without errors in background"):
+        fio_pid = good_core_fio.run_in_background()
+
+    fio_errors = 0
+    with TestRun.step(
+        "Run fio on error core and check if error count matches with cache stats"
```
Comment (mmichal10) [inline test/functional/tests/fault/test_primary_device_error.py]: Actually you don't check cache stats in this step

File: test/functional/tests/fault/test_primary_device_error.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    ):
+        fio_errors = error_core_fio.run()[0].total_errors()
+
+        if fio_errors == 0:
+            TestRun.fail("No I/O ended with error!")
+
+    with TestRun.step("Check error statistics on error core"):
```
Comment (mmichal10) [inline test/functional/tests/fault/test_primary_device_error.py]: Shouldn't this step and the previous one be in reversed order?

File: test/functional/tests/fault/test_primary_device_error.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+        description: |
+          Perform I/O on two exported objects created using error and non-error device.
+          Validate CAS that stats counting is consistent with OS reporting.
+          Also, check if normal I/O is uninterrupted and no DC occurs on any of the
+          core devices.
+        pass_criteria:
+          - I/O error count in FIO and in cache statistics match
+          - Positively passed fio verify on both core devices
+    """
+    cache, error_core, error_core_device, good_core, good_core_device = prepare_configuration(
```
Comment (mmichal10) [inline test/functional/tests/fault/test_primary_device_error.py]: Isn't error_core_device and good_core_device redundant? You can retrieve those values from error_core and good_core.

File: test/functional/tests/fault/test_primary_device_error.py (hunk 6 above)
Comment (imjfckm) [reply test/functional/tests/fault/test_primary_device_error.py]: What do you mean? Switch "Check error~~" and "Run fio~~"? There won't be any errors on the device before we run fio

File: test/functional/tests/fault/test_primary_device_error.py (hunk 5 above)
Comment (imjfckm) [reply test/functional/tests/fault/test_primary_device_error.py]: Fixed

File: test/functional/tests/fault/test_primary_device_error.py (hunk 7 above)
Comment (imjfckm) [reply test/functional/tests/fault/test_primary_device_error.py]: True that, fixed

File: test/functional/tests/fault/test_primary_device_error.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        .io_engine(IoEngine.libaio)
+        .size(error_core.size)
+        .block_size(cache_line_size)
+        .target(error_core)
+        .read_write(io_dir)
+        .continue_on_error(ErrorFilter.io)
+        .direct()
+    )
+    if io_dir == ReadWrite.randwrite:
+        error_core_fio = error_core_fio.verify_pattern().verify(VerifyMethod.pattern)
```
Comment (robertbaldyga) [inline test/functional/tests/fault/test_primary_device_error.py]: This assignment is unnecessary.

Comment (imjfckm) [reply test/functional/tests/fault/test_primary_device_error.py]: Is it bad tho? For me it's easier to read that way

File: test/functional/tests/fault/test_primary_device_error.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+        if fio_errors == 0:
+            TestRun.fail("No I/O ended with error!")
+
+    with TestRun.step("Check error statistics on error core"):
+        stats = cache.get_statistics()
+
+        core_errors_in_cache = stats.error_stats.core.total
+        if fio_errors != core_errors_in_cache:
+            TestRun.fail(
+                f"Core errors in cache stats({core_errors_in_cache}) should equal to number of"
```
Comment (Deixx) [inline test/functional/tests/fault/test_primary_device_error.py]: should be equal

File: test/functional/tests/fault/test_primary_device_error.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+    if io_dir == ReadWrite.randwrite:
+        with TestRun.step("Verify error core device contents"):
+            error_core_fio.target(error_core.core_device).verify_only().run()
+
+    with TestRun.step("Verify good core device contents"):
+        good_core_fio.target(good_core.core_device).verify_only().run()
+
+
+def prepare_configuration(cache_mode, cache_line_size):
```
Comment (Deixx) [inline test/functional/tests/fault/test_primary_device_error.py]: I thought that for the purpose of parsing test steps, we weren't going to enclose steps within methods

File: test/functional/tests/fault/test_primary_device_error.py (hunk 8 above)
Comment (Deixx) [reply test/functional/tests/fault/test_primary_device_error.py]: It's not incorrect, but it isn't necessary and readability impact seems negligible ;) That's just my opinion - I won't fight for it ;)

Comment (imjfckm) [reply test/functional/tests/fault/test_primary_device_error.py]: Neither will I! Changed :)

File: test/functional/tests/fault/test_primary_device_error.py (hunk 9 above)
Comment (imjfckm) [reply test/functional/tests/fault/test_primary_device_error.py]: Fixed

File: test/functional/tests/fault/test_primary_device_error.py (hunk 10 above)
Comment (imjfckm) [reply test/functional/tests/fault/test_primary_device_error.py]: Fixed, also removed conditional step to make it easier

File: test/functional/tests/fault/test_primary_device_error.py [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+from test_tools.fio.fio_param import ReadWrite, IoEngine, ErrorFilter, VerifyMethod
+from test_tools.device_mapper import ErrorDevice, DmTable
+from core.test_run import TestRun
+from api.cas import casadm
+from api.cas.cache_config import (
+    CacheMode,
+    CacheLineSize,
+    SeqCutOffPolicy,
+    CleaningPolicy,
+    CacheStatus,
```
Comment (Deixx) [inline test/functional/tests/fault/test_primary_device_error.py]: not used

File: test/functional/tests/fault/test_primary_device_error.py [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+)
+from storage_devices.disk import DiskTypeSet, DiskType, DiskTypeLowerThan
+from storage_devices.device import Device
```
Comment (Deixx) [inline test/functional/tests/fault/test_primary_device_error.py]: not used

File: test/functional/tests/fault/test_primary_device_error.py [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+from test_utils.os_utils import Udev, sync
```
Comment (Deixx) [inline test/functional/tests/fault/test_primary_device_error.py]: sync not used

File: test/functional/tests/fault/test_primary_device_error.py [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+from test_utils.size import Size, Unit
+from test_utils.filesystem.file import File
```
Comment (Deixx) [inline test/functional/tests/fault/test_primary_device_error.py]: not used

File: test/functional/tests/fault/test_primary_device_error.py [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+        .continue_on_error(ErrorFilter.io)
+        .direct()
+    )
+    if io_dir == ReadWrite.randwrite:
+        error_core_fio.verify_pattern().verify(VerifyMethod.pattern)
+
+    with TestRun.step("Run fio on core without errors in background"):
+        fio_pid = good_core_fio.run_in_background()
+
+    fio_errors = 0
```
Comment (Deixx) [inline test/functional/tests/fault/test_primary_device_error.py]: not used

File: test/functional/tests/fault/test_primary_device_error.py (hunk 11 above)
Comment (imjfckm) [reply test/functional/tests/fault/test_primary_device_error.py]: Fixed

File: test/functional/tests/fault/test_primary_device_error.py (hunk 12 above)
Comment (imjfckm) [reply test/functional/tests/fault/test_primary_device_error.py]: Fixed

File: test/functional/tests/fault/test_primary_device_error.py (hunk 13 above)
Comment (imjfckm) [reply test/functional/tests/fault/test_primary_device_error.py]: Fixed

File: test/functional/tests/fault/test_primary_device_error.py (hunk 14 above)
Comment (imjfckm) [reply test/functional/tests/fault/test_primary_device_error.py]: Fixed

File: test/functional/tests/fault/test_primary_device_error.py (hunk 15 above)
Comment (imjfckm) [reply test/functional/tests/fault/test_primary_device_error.py]: Fixed

**Pull request merged**